pkg/splunk/enterprise/monitoringconsole.go

```diff
@@ -0,0 +1,342 @@
+	addNewURLs := true
+
+	if cr.GetObjectMeta().GetDeletionTimestamp() != nil {
+		addNewURLs = false
+	}
+
+	_, err = ApplyMonitoringConsoleEnvConfigMap(client, cr.GetNamespace(), cr.GetName(), extraEnv, addNewURLs)
+	if err != nil {
+		return err
+	}
+
+	//configMapHash to trigger the configMap change in monitoring console deployment
+	configMapHash, err := getConfigMapHash(client, cr.GetNamespace())
+	if err != nil {
+		return err
+	}
+
+	//configMapHash should never be nil, just adding extra check
+	if configMapHash == "" {
+		return nil
+	}
+
+	deploymentMC, err := getMonitoringConsoleDeployment(cr, &spec, SplunkMonitoringConsole, 1, configMapHash, secretName)
+	if err != nil {
+		return err
+	}
+
+	_, err = splctrl.ApplyDeployment(client, deploymentMC)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// GetMonitoringConsoleDeployment returns a Kubernetes Deployment object for Splunk Enterprise monitoring console instance.
+func getMonitoringConsoleDeployment(cr splcommon.MetaObject, spec *enterprisev1.CommonSplunkSpec, instanceType InstanceType, replicas int32, configMapHash string, secretName string) (*appsv1.Deployment, error) {
+	var partOfIdentifier string
+	ports := splcommon.SortContainerPorts(getSplunkContainerPorts(SplunkMonitoringConsole))
+	annotations := splcommon.GetIstioAnnotations(ports)
+	//using Namespace here so that with every CR the name should remain same Ex- splunk-<namespace>-monitoring-console
+	selectLabels := getSplunkLabels(cr.GetNamespace(), instanceType, partOfIdentifier)
+	affinity := splcommon.AppendPodAntiAffinity(&spec.Affinity, cr.GetNamespace(), instanceType.ToString())
+	configMap := GetSplunkMonitoringconsoleConfigMapName(cr.GetNamespace(), SplunkMonitoringConsole)
+
+	// start with same labels as selector; note that this object gets modified by splcommon.AppendParentMeta()
+	labels := make(map[string]string)
+	for k, v := range selectLabels {
+		labels[k] = v
+	}
+
+	//create deployment configuration
+	deploymentMC := &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetSplunkMonitoringConsoleDeploymentName(instanceType, cr.GetNamespace()),
+			Namespace: cr.GetNamespace(),
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: selectLabels,
+			},
+			Replicas: &replicas,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      labels,
+					Annotations: annotations,
+				},
+				Spec: corev1.PodSpec{
+					Affinity:      affinity,
+					Tolerations:   spec.Tolerations,
+					SchedulerName: spec.SchedulerName,
+					Containers: []corev1.Container{
+						{
+							Image:           spec.Image,
+							ImagePullPolicy: corev1.PullPolicy(spec.ImagePullPolicy),
+							Name:            "splunk",
+							Ports:           ports,
+							EnvFrom: []corev1.EnvFromSource{
```

Comment:
Would you please explain again why we are setting env variables via a configmap, versus just embedding the key/values here within the pod spec?

Reply:
For MC, it needs to have URLs of all the splunk instances running in that namespace. If we just embedded the URLs directly as env variables, one CR will not have any knowledge about the other URLs from other statefulset or custom resource. So we have separate configMap where we add/delete all the URLs for that namespace